improver/snow_fraction.py

```diff
@@ -0,0 +1,164 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing the SnowFraction class."""
+
+
+import iris
+
+from improver import PostProcessingPlugin
+from improver.metadata.utilities import (
+    create_new_diagnostic_cube,
+    generate_mandatory_attributes,
+)
+from improver.utilities.cube_checker import spatial_coords_match
+from improver.utilities.interpolation import interpolate_missing_data
+
+
+class SnowFraction(PostProcessingPlugin):
+    """
+    Calculates a snow-fraction field from fields of snow and rain (rate or
+    accumulation). Where no precipitation is present, the data are filled in from
+    the nearest precipitating point.
+    """
+
+    def _get_input_cubes(self, input_cubes):
+        """
+        Separates out the rain and snow cubes from the input list and checks that
+            * No other cubes are present
+            * Cubes represent the same time quantity (instantaneous or accumulation length)
+            * Cubes have compatible units
+            * Cubes have same dimensions
+
+        Args:
+            input_cubes (iris.cube.CubeList):
+                Contains exactly two cubes, one of rain and one of snow. Both either
+                rates or accumulations of the same length and of compatible units.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError:
+                If any of the criteria above are not met.
+
+        """
+        if len(input_cubes) != 2:
+            raise ValueError(
+                f"Expected exactly two input cubes, found {len(input_cubes)}"
+            )
+        rain_name, snow_name = self._get_input_cube_names(input_cubes)
+        self.rain = input_cubes.extract(rain_name).merge_cube()
+        self.snow = input_cubes.extract(snow_name).merge_cube()
+        self.snow.convert_units(self.rain.units)
+        if not spatial_coords_match(self.rain, self.snow):
+            raise ValueError("Rain and Snow cubes are not on the same grid")
+        if not self.rain.coord("time") == self.snow.coord("time"):
+            raise ValueError("Rain and Snow cubes do not have the same time coord")
+
+    def _get_input_cube_names(self, input_cubes):
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6311273)